### PR TITLE
mgr/dashboard: Prevent table items from getting selected while expanding

### PR DESCRIPTION
**Pull request by nizamial09**
While clicking on the expand button inside a table, the entire row is getting selected. This commit prevent that from happening.

captured (1)

Fixes: https://tracker.ceph.com/issues/47376
Signed-off-by: Nizamudeen A <nia@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by tspmelo — CHANGES_REQUESTED** [review]
Please add a new unit test to check if the `selection` is not changed when expanding a different row.

**Review by epuertat — APPROVED** [review]
Once addressed @tspmelo's comments, I'm ok with this PR! Thanks @nizamial09!

**Comment by nizamial09** [comment]
> Please add a new unit test to check if the `selection` is not changed when expanding a different row.

Added the unit test. Please take a look, thanks. @tspmelo

**Review by avanthakkar — APPROVED** [review]
Tested it locally, LGTM!

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts [hunk 1]
```diff
@@ -686,18 +686,28 @@ describe('TableComponent', () => {
...
     });
+
+    it('should not change selection when expanding different row', () => {
+      spyOn(component, 'onSelect').and.callThrough();
+      component.data = createFakeData(10);
+      component.selection.selected = [_.clone(component.data[1])];
+      component.expanded = _.clone(component.data[2]);
+      component.toggleExpandRow(component.expanded, false, new Event('click'));
+      expect(component.selection.selected).toEqual([{ a: 1, b: 10, c: true }]);
+      expect(component.expanded).toEqual({ a: 2, b: 20, c: false });
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts:710]: I think using the following is more explicit and removes some logic that is not needed.
```suggestion
      expect(component.selection.selected).toEqual([]);
      expect(component.expanded).toEqual(component.data[1]);
      component.selection.selected = [component.data[2]];
      component.toggleExpandRow(component.data[3], false, new Event('click'));
      expect(component.selection.selected).toEqual([component.data[2]]);
      expect(component.expanded).toEqual(component.data[3]);
```

**Review by tspmelo — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts]: Thank you @tspmelo Updated the PR as you suggested.

**Comment by nizamial09** [comment]
@avanthakkar Just saw your suggestion. The suggestion was edited into my comment so I missed it before. Anyway I've updated the description with the new gif. Thanks

**Review by tspmelo — APPROVED** [review]

**Comment by avanthakkar** [comment]
> @avanthakkar Just saw your suggestion. The suggestion was edited into my comment so I missed it before. Anyway I've updated the description with the new gif. Thanks

Yeah, I just saw that, my mistake. Thanks!

**Review by votdev — CHANGES_REQUESTED** [review]
I think you should check if e2e test aren't broken by this PR.
```
  Configuration page
    breadcrumb test
      ✓ should open and show breadcrumb (2709ms)
    fields check
      1) should verify that selected footer increases when an entry is clicked
      ✓ should check that details table opens (w/o tab header) (1077ms)
    edit configuration test
      ✓ should click and edit a configuration and results should appear in the table (9299ms)
      ✓ should show only modified configurations (1439ms)
      ✓ should hide all modified configurations (1457ms)


  5 passing (27s)
  1 failing

  1) Configuration page
       fields check
         should verify that selected footer increases when an entry is clicked:
     AssertionError: expected 0 to equal 1
      at Context.eval (https://bionic-huge--734e7c1f-b7d8-454f-a114-e1abf5a3d023:41140/__cypress/tests?p=cypress/integration/cluster/configuration.e2e-spec.ts:21:53)
```

**Comment by nizamial09** [comment]
> I think you should check if e2e test aren't broken by this PR.
> 
> ```
>   Configuration page
>     breadcrumb test
>       ✓ should open and show breadcrumb (2709ms)
>     fields check
>       1) should verify that selected footer increases when an entry is clicked
>       ✓ should check that details table opens (w/o tab header) (1077ms)
>     edit configuration test
>       ✓ should click and edit a configuration and results should appear in the table (9299ms)
>       ✓ should show only modified configurations (1439ms)
>       ✓ should hide all modified configurations (1457ms)
> 
> 
>   5 passing (27s)
>   1 failing
> 
>   1) Configuration page
>        fields check
>          should verify that selected footer increases when an entry is clicked:
>      AssertionError: expected 0 to equal 1
>       at Context.eval (https://bionic-huge--734e7c1f-b7d8-454f-a114-e1abf5a3d023:41140/__cypress/tests?p=cypress/integration/cluster/configuration.e2e-spec.ts:21:53)
> ```

Oh yeah, thanks. Let me check this locally.

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
@votdev I removed certain lines from the e2e files like `should verify that selected footer increases when an entry is clicked`. IMO this feels like a unit test and similar unit test is already covered in this PR (row getting selected).

**Review by votdev — APPROVED** [review]

**Pull request merged**